### PR TITLE
[AOTInductor] Support freezing model in aoti

### LINKED ISSUE #114450: [AOTInductor] Need support to export freezing model
### 🚀 The feature, motivation and pitch

AOTInductor not support freezing because can not deal with opaque tensor.
I've come across a need to deploy a quantized model, which means I need to combine aoit and pt2e quant. The first obstacle encountered is inductor freezing. Open inductor config freezing before aot_compile will see the error: "NotImplementedError: Cannot access storage of OpaqueTensorImpl".

### Alternatives

_No response_

### Additional context

_No response_

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by imzhuhl**
Fixes #114450

This commit supports aotinductor processing opaque tensor, so we can enable freezing in aoti. Need help from the community to review and see if this is a suitable way.

Test on cpu only device:
```bash
python test_aot_inductor.py -k AOTInductorTestNonABICompatibleCpu.test_freezing_non_abi_compatible_cpu 
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/csrc/inductor/aoti_torch/c/shim.h [hunk 1]
```diff
@@ -170,6 +170,7 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch_create_tensor_from_blob(
     const int64_t* strides_ptr,
     int64_t storage_offset,
     int32_t dtype,
+    int8_t layout,
```
Comment (desertfire) [inline torch/csrc/inductor/aoti_torch/c/shim.h:173]: All the APIs in this file is supposed to be ABI-stable. If you need to add a parameter like this, you will have to create a v2 version of this function.

File: test/inductor/test_aot_inductor.py [hunk 2]
```diff
@@ -1454,5 +1453,5 @@ class AOTInductorTestNonABICompatibleCuda(TestCase):
     from torch._dynamo.test_case import run_tests
 
     # cpp_extension N/A in fbcode
-    if HAS_CUDA and not TEST_WITH_ROCM:
+    if not TEST_WITH_ROCM:
```
Comment (desertfire) [inline test/inductor/test_aot_inductor.py:1456]: You need to make sure all the cuda tests are decorated with `@requires_cuda` if you want to remove this.

File: torch/csrc/inductor/aoti_torch/opaque_tensor.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <ATen/Tensor.h>
+
+namespace torch {
+namespace aot_inductor {
+
+at::Tensor mkldnn_tensor_from_data_ptr(
```
Comment (desertfire) [inline torch/csrc/inductor/aoti_torch/opaque_tensor.h:8]: If `mkldnn_tensor_from_data_ptr` is only used by `aoti_torch_create_tensor_from_blob`, it's better to keep it as a utility function in shim_common.cpp. Files in `atoi_torch` are a set of limited interfaces used by the generated `model.so`, so adding a generic torch function here can cause confusion and people may use it by mistake.

**Review by desertfire — CHANGES_REQUESTED** [review]
First of all, thanks for contributing to AOTInductor! I left some comments about the aoti_torch interfaces, and I will leave MKLDNN-related part to @jgong5 and @chunyuan-w

File: torch/csrc/inductor/aoti_torch/opaque_tensor.cpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+#if AT_MKLDNN_ENABLED()
+
+at::Tensor mkldnn_tensor_from_data_ptr(
+    void* data_ptr,
+    at::IntArrayRef dims,
+    at::ScalarType dtype,
+    at::Device device) {
+  auto a = ideep::tensor({dims.vec(), ideep::tensor::data_type::s8}, data_ptr);
+  return at::native::new_with_itensor_mkldnn(std::move(a), dtype, device);
+}
```
Comment (jgong5) [inline torch/csrc/inductor/aoti_torch/opaque_tensor.cpp:21]: Oh, I don't think this is a right way of serializing/deserializing mkldnn tensors. The MKLDNN tensor contains both the memory buffer (the `data_ptr` points to) and the metadata that describes the layout of the memory buffer. When the model is frozen, the weight tensors are with blocked layout which is described with the metadata. So, only serializing/deserializing the memory buffer cannot restore all the information held in the MKLDNN tensor. oneDNN folks are adding the serialization support for MKLDNN tensors per our request. I would suggest we wait for the support ready before adding the support for frozen models.

**Review by jgong5 — CHANGES_REQUESTED** [review]
It doesn't seem the right way of serializing/deserializing the MKLDNN tensors. Inline comments added for details.

**Comment by imzhuhl** [comment]
> First of all, thanks for contributing to AOTInductor! I left some comments about the aoti_torch interfaces, and I will leave MKLDNN-related part to @jgong5 and @chunyuan-w

Thanks for your review. I will make changes. 
@jgong5 is right, mkldnn tensor serialization and deserialization is not that simple, it requires a deeper understanding of the oneDNN prepack mechanism, I need some time to figure this out. If it is difficult to complete in this pr, waiting for support from oneDNN team may be the best option.

File: torch/csrc/inductor/aoti_torch/opaque_tensor.cpp (hunk 4 above)
Comment (jgong5) [reply torch/csrc/inductor/aoti_torch/opaque_tensor.cpp:21]: We are working on a POC and will tag you when the PR is ready. cc @chunyuan-w

**Comment by eellison** [comment]
Clearing review queue please re-request when needed

**Pull request closed without merging**